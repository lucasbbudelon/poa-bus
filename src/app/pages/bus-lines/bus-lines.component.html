<section class="m-3">
  <h1>POA-Bus</h1>
  <p class="lead">Sistema integrado ao <a href="http://datapoa.com.br/group/about/mobilidade"
      target="_blanck">DataPoa</a>
    para consulta de linhas de ônibus e linhas de lotação.</p>

  <form [formGroup]="formGroup" *ngIf="data">

    <div class="form-group form-inline">
      Buscar por: <input class="form-control ml-2" type="text" formControlName="search" />
    </div>

    <table class="table table-striped">
      <thead>
        <tr>
          <th scope="col" style="width: 5%;">#</th>
          <th scope="col" style="width: 10%;">Linha</th>
          <th scope="col" style="width: 10%;">Tipo</th>
          <th scope="col" style="width: 50%;">Nome <i class="fas fa-sort-alpha-up"></i></th>
        </tr>
      </thead>
      <tbody>
        <tr *ngFor="let busLine of data.items; let i = index">
          <th scope="row">
            {{ (i+1) + (currentPage > 1 ? (currentPage -1) * pageSize.value : 0) }}
          </th>
          <td>
            {{busLine.id}} # {{busLine.code}}
          </td>
          <td [ngSwitch]="busLine.type">
            <ng-container *ngSwitchCase="busLineType.Bus">
              <i class="fa fa-bus"></i> Ônibus
            </ng-container>
            <ng-container *ngSwitchCase="busLineType.Stocking">
              <i class="fas fa-shuttle-van"></i> Lotação
            </ng-container>
          </td>
          <td>
            {{busLine.name}}
          </td>
        </tr>
      </tbody>
    </table>

    <div class="d-flex justify-content-between mt-4">
      <ngb-pagination [collectionSize]="data.totalItems" [(page)]="currentPage" [pageSize]="pageSize.value"
        (pageChange)="loadNextPage()" [maxSize]="paginationMaxSize" [boundaryLinks]="false">
      </ngb-pagination>

      <select class="custom-select w-auto" formControlName="pageSize">
        <option *ngFor="let value of itemsPerPage" [ngValue]="value">{{value}} itens por página</option>
      </select>
    </div>
  </form>
</section>
