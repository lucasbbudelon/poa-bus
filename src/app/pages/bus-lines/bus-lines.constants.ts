import { NgbModalOptions } from '@ng-bootstrap/ng-bootstrap';

export const ITEMS_PER_PAGE = [5, 10, 15, 20, 25, 30];
export const ITEMS_PER_PAGE_DEFAULT = 10;
export const PAGINATION_MAX_SIZE = 10;
export const MODAL_ITINERARY_OPTIONS: NgbModalOptions = {
    scrollable: true,
    size: 'lg'
};
