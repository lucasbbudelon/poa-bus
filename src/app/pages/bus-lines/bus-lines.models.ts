export interface BusLine {
  id: string;
  code: string;
  name: string;
  type: BusLineType;
}

export interface BusLinePage {
  current?: number;
  limit?: number;
}

export interface BusLineData {
  page: BusLinePage;
  items: BusLine[];
  totalItems: number;
}

export enum BusLineType {
  Bus = 'o',
  Stocking = 'l',
}

export interface BusLineItinerary {
  busLine: BusLine;
  locations: Location[];
}

export interface BusLineItineraryLocation {
  order: number;
  lat: string;
  lng: string;
  urlMap: string;
}
