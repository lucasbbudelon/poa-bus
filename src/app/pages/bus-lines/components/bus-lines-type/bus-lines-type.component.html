    <span [ngSwitch]="type">
      <ng-container *ngSwitchCase="busLineType.Bus">
        <i class="fa fa-bus"></i> Ônibus
      </ng-container>
      <ng-container *ngSwitchCase="busLineType.Stocking">
        <i class="fas fa-shuttle-van"></i> Lotação
      </ng-container>
    </span>
