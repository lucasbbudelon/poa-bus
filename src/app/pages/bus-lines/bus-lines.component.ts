import { Component, OnDestroy, OnInit } from '@angular/core';
import { AbstractControl, FormControl, FormGroup } from '@angular/forms';
import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
import { Subscription } from 'rxjs';
import { finalize, switchMap, tap } from 'rxjs/operators';

import { ITEMS_PER_PAGE, ITEMS_PER_PAGE_DEFAULT, MODAL_ITINERARY_OPTIONS, PAGINATION_MAX_SIZE } from './bus-lines.constants';
import { BusLine, BusLineData, BusLinePage } from './bus-lines.models';
import { BusLinesService } from './bus-lines.service';
import { BusLinesItineraryComponent } from './components/bus-lines-itinerary/bus-lines-itinerary.component';

@Component({
  selector: 'app-bus-lines',
  templateUrl: './bus-lines.component.html',
  styleUrls: ['./bus-lines.component.scss'],
})
export class BusLinesComponent implements OnInit, OnDestroy {
  get search(): AbstractControl {
    return this.formGroup.get('search');
  }
  get pageSize(): AbstractControl {
    return this.formGroup.get('pageSize');
  }
  get itemsPerPage() {
    return ITEMS_PER_PAGE;
  }
  get paginationMaxSize() {
    return PAGINATION_MAX_SIZE;
  }
  get firstPage(): BusLinePage {
    return { current: 1, limit: this.pageSize.value };
  }
  get nextPage(): BusLinePage {
    return { current: this.currentPage++, limit: this.pageSize.value };
  }

  public formGroup: FormGroup;
  public data: BusLineData;
  public currentPage: number;

  private subscriptions: Subscription;

  constructor(
    private busLinesService: BusLinesService,
    private modalService: NgbModal
  ) { }

  ngOnInit() {
    this.buildForm();
    this.loadFirstPage();
  }

  ngOnDestroy() {
    if (this.subscriptions && !this.subscriptions.closed) {
      this.subscriptions.unsubscribe();
    }
  }

  buildForm() {
    this.formGroup = new FormGroup({
      search: new FormControl(null),
      pageSize: new FormControl(ITEMS_PER_PAGE_DEFAULT),
    });

    this.subscriptions = new Subscription();
    this.subscribeToSearchChanges();
    this.subscribeToPageSizeChanges();
  }

  subscribeToSearchChanges() {
    const searchValueChanges$ = this.search.valueChanges
      .pipe(
        switchMap((filter) => this.busLinesService
          .getBusLines(this.firstPage, filter))
      )
      .subscribe((data) => this.data = data);

    this.subscriptions.add(searchValueChanges$);
  }

  subscribeToPageSizeChanges() {
    const pageSizeValueChanges$ = this.pageSize.valueChanges
      .pipe(
        switchMap((value) => this.busLinesService
          .getBusLines({ current: this.currentPage, limit: value }))
      )
      .subscribe((data) => this.data = data);

    this.subscriptions.add(pageSizeValueChanges$);
  }

  loadFirstPage() {
    this.busLinesService
      .getBusLines(this.firstPage)
      .pipe(
        tap((data) => this.data = data),
        tap((data) => this.currentPage = data.page.current),
      )
      .subscribe();
  }

  loadNextPage() {
    this.busLinesService
      .getBusLines(this.nextPage)
      .pipe(
        tap((data) => this.data = data),
        tap((data) => this.currentPage = data.page.current),
      )
      .subscribe();
  }

  openItinerary(busLine: BusLine) {
    this.busLinesService
      .getItinerary(busLine)
      .subscribe((itinerary) => {
        const busLinesItineraryComponent = this.modalService.open(BusLinesItineraryComponent, MODAL_ITINERARY_OPTIONS);
        busLinesItineraryComponent.componentInstance.itinerary = itinerary;
      });
  }
}
