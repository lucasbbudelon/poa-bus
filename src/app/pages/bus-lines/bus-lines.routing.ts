import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import { BusLinesComponent } from './bus-lines.component';

const routes: Routes = [
  {
    path: '',
    component: BusLinesComponent,
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class BusLinesRouting {}
