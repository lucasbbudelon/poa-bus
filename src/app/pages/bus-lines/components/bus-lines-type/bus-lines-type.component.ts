import { Component, Input } from '@angular/core';

import { BusLineType } from '../../bus-lines.models';

@Component({
  selector: 'app-bus-lines-type',
  templateUrl: './bus-lines-type.component.html',
  styleUrls: ['./bus-lines-type.component.scss']
})
export class BusLinesTypeComponent {
  @Input() type: BusLineType;
  get busLineType() { return BusLineType; }
  constructor() { }
}
