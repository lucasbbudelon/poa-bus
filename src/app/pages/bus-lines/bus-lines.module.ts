import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { NgbAccordionModule, NgbPaginationModule } from '@ng-bootstrap/ng-bootstrap';

import { BusLinesComponent } from './bus-lines.component';
import { BusLinesRouting } from './bus-lines.routing';
import { BusLinesItineraryComponent } from './components/bus-lines-itinerary/bus-lines-itinerary.component';
import { BusLinesTypeComponent } from './components/bus-lines-type/bus-lines-type.component';

@NgModule({
  declarations: [
    BusLinesComponent,
    BusLinesItineraryComponent,
    BusLinesTypeComponent
  ],
  imports: [
    CommonModule,
    BusLinesRouting,
    FormsModule,
    ReactiveFormsModule,
    NgbPaginationModule,
    NgbAccordionModule
  ],
})
export class BusLinesModule { }
