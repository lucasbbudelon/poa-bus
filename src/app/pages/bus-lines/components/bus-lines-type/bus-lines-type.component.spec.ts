import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { BusLinesTypeComponent } from './bus-lines-type.component';
import { BusLineType } from '../../bus-lines.models';

describe('BusLinesTypeComponent', () => {
  let component: BusLinesTypeComponent;
  let fixture: ComponentFixture<BusLinesTypeComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [BusLinesTypeComponent]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(BusLinesTypeComponent);
    component = fixture.componentInstance;
    Object.defineProperty(component, 'busLineType', {
      get: jest.fn().mockReturnValue(BusLineType),
    });
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('when type=Bus', () => {
    component.type = BusLineType.Bus;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('when type=Stocking', () => {
    component.type = BusLineType.Stocking;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
