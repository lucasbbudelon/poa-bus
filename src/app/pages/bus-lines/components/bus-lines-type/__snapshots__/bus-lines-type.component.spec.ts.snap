// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusLinesTypeComponent when type=Bus 1`] = `
<app-bus-lines-type
  type={[Function String]}
>
  <span>
    
    
    <i
      class="fa fa-bus"
    />
     Ônibus 
    
  </span>
</app-bus-lines-type>
`;

exports[`BusLinesTypeComponent when type=Stocking 1`] = `
<app-bus-lines-type
  type={[Function String]}
>
  <span>
    
    
    
    <i
      class="fas fa-shuttle-van"
    />
     Lotação 
  </span>
</app-bus-lines-type>
`;
